datafusion/optimizer/src/type_coercion.rs

```diff
@@ -119,6 +134,41 @@ impl ExprRewriter for TypeCoercionRewriter {
 
     fn mutate(&mut self, expr: Expr) -> Result<Expr> {
         match expr {
+            Expr::ScalarSubquery(Subquery { subquery }) => {
+                let mut optimizer_config = OptimizerConfig::new();
+                let new_plan =
+                    optimize_internal(&self.schema, &subquery, &mut optimizer_config)?;
+                Ok(Expr::ScalarSubquery(Subquery::new(new_plan)))
+            }
+            Expr::Exists { subquery, negated } => {
+                let mut optimizer_config = OptimizerConfig::new();
+                let new_plan = optimize_internal(
```

Comment:
I don't understand why the outer query's schema needs to be passed while optimizing the subquery. I would expect that we would recursively optimize the `subquery`. The subquery doesn't have the same schema as its containing query 🤔

Reply:
As in I think this should be more like:

```rust
let optimizer = TypeCoercion::new();
let new_plan = optimizer.optimize(&subquery.subquery, &mut optimizer_config)
```

Reply:
I also want this, but there are some correlated subquery, like the comments.
I have added the comments about the usage of `external_schema` schema
    // merge the outer schema for correlated subqueries
    // like case:
    // select t2.c2 from t1 where t1.c1 in (select t2.c1 from t2 where t2.c2=t1.c3)

For the subquery `select t2.c1 from t2 where t2.c2= t1.c3`, if we don't know the schema about the `t1`, we can get the `data type` for `t1.c3` and can't do the type coercion for `t2.c2=t1.c3`.

This is way that I find, do you @andygrove @alamb  have any thoughts about this?

Reply:
I think I would need to try it myself